packages/flutter_tools/lib/src/commands/upgrade.dart

```diff
@@ -115,8 +115,8 @@ class UpgradeCommandRunner {
     } else if (verifyOnly) {
       globals.printStatus('A new version of Flutter is available on channel ${flutterVersion.channel}\n');
       // TODO(fujino): use a [FlutterVersion] once that class supports arbitrary revisions.
-      globals.printStatus('The latest revision: $upstreamRevision', emphasis: true);
-      globals.printStatus('Your current version: ${flutterVersion.frameworkRevision}\n');
+      globals.printStatus('The latest version: ${gitTagVersion.frameworkVersionFor(upstreamRevision)}', emphasis: true);
```

Comment:
So something like `git describe --tags --abbrev=0` would suffice? Seems working to me running in my cloned repo.

Reply:
> So something like `git describe --tags --abbrev=0` would suffice? Seems working to me running in my cloned repo.

Unfortunately, if the commit in question is not a tagged commit, that will report the **nearest** tag. However, if a commit is 7 commits past version 1.2.3, it is not accurate to report that it as 1.2.3. What we need is to support constructing a `GitTagVersion` for a revision other than `HEAD`. Then I would update the function `Future<String> fetchRemoteRevision()` to instead be `Future<FlutterVersion> fetchLatestVersion()`.

Reply:
The problem is that `GitTagVersion.determine()` has this line:

```
    final List<String> tags = _runGit(
      'git tag --points-at HEAD', processUtils, workingDirectory).trim().split('\n');
```

which hard-codes `HEAD`.

Reply:
And then also the fallback at the end implicitly checks HEAD:

```
    // If we're not currently on a tag, use git describe to find the most
    // recent tag and number of commits past.
    return parse(
      _runGit(
        'git describe --match *.*.* --first-parent --long --tags',
        processUtils,
        workingDirectory,
      )
    );
```

Reply:
>What we need is to support constructing a GitTagVersion for a revision other than HEAD.

Changing that would break most of the tests in `version_test.dart`.

>Then I would update the function Future<String> fetchRemoteRevision() to instead be Future<FlutterVersion> fetchLatestVersion().

Changing this would break tests at `upgrade_test.dart` as well.

I think this PR might be a breaking change.